### ISSUE TITLE
bug: executing JavaScript on iOS failes due to "Recursive object cannot be transferred"

### ISSUE THREAD

**Issue by atmaxinger** [appium/appium#20475]
### Do I have the most recent component updates?

- [X] I use the most recent available driver/plugin and server versions

### Is the component officially supported by the Appium team?

- [X] I have verified the component repository is present under the Appium organization in GitHub

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When executing JavaScript that returns an array or object tree with the same object multiple times (non-recursively), the execution fails with the following error:

```
ERROR webdriver: Request failed with status 500 due to javascript error: Recursive object cannot be transferred
```

For example, the following JavaScript would fail:

```javascript
let x={hello: 'world'};
let a=[x,x];
return a;
```

### Expected Behavior

The result is returned correctly.

### Minimal Reproducible Example

```javascript
const {remote} = require('webdriverio');

const capabilities = {
  platformName: 'iOS',
  'appium:automationName': 'XCUITest',
  'appium:udid': '<DEVICE UDID>',
  'appium:browserName': 'Safari',
};

const wdOpts = {
  hostname: 'localhost',
  port: 4723,
  logLevel: 'info',
  capabilities,
};

async function runTest() {
  const driver = await remote(wdOpts);
  let result = driver.executeScript("let x={hello: 'world'};let a=[x,x];return a;", []);
}

runTest().catch(console.error);
```

### Further Information

Upon some investigation, this is because of this code in the JSONWP implementation in Selenium: https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/atoms/inject.js#L123-L126

They've implemented a naive and wrong heuristic to check for recursive elements.

Is there a way to use the XCUITest driver without JSONWP?

**Comment by mykola-mokhnach** [comment]
> Upon some investigation, this is because of this code in the JSONWP implementation in Selenium: https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/atoms/inject.js#L123-L126

> They've implemented a naive and wrong heuristic to check for recursive elements.

> Is there a way to use the XCUITest driver without JSONWP?

Unfortunately currently there is no way to call javascript differently (e.g. without using selenium atoms).

There are two possible options there:

- Report/fix selenium atoms and bump them in the appium remote debugger component.
- Change the implementation of appium remote debugger component to not use atoms and replace them with some alternative. Unfortunately we don't have enough expertise in the team to make that possible now. Maybe someone could help there.

**Comment by mykola-mokhnach** [comment]
@ahalbrock Could you please take a look at this issue?

**Comment by ahalbrock** [comment]
I agree that this causes an error to occur but I'm unsure what you are going for here.  Executing JavaScript through the driver is not supposed to "physically" transport something from between your client code (whether running in java, objective c, or JavaScript) and the iOS app, it recreates things on the other side, so basically serializing and deserializing the objects.  This would break any recursive definitions anyway.  I do see that running this case on Android (utilizing however the UIAutomator2 driver does this stuff) does not cause error however.  I can only assume that your real scenario contains code that at least sometimes produces objects that are recursively defined like this. 

I can take a look around the code to see how this works, but if you want to get around this you could simply wrap a JSON parse/stringify around your returned object and you will get something indistinguishable from what you'd get if the error wasn't thrown. So using this instead:

`
let result = driver.executeScript("let x={hello: 'world'};let a=[x,x];return JSON.parse(JSON.stringify(a));", []);
`

**Comment by mykola-mokhnach** [comment]
Closed as third party issue
